Breaking: Deprecate Reach-UI based modal and add Ariakit-based modal

frankieyan opened the pull request:
<!--
Include a link to related issues, or more importantly, any issue this may close.
-->

Closes https://github.com/Doist/reactist/issues/679

## Short description

This re-implements the modal with Ariakit. It tries to match the previous implementation as close as possible, both API and markup wise, so we don't leverage Ariakit's default modal backdrop.

## Test plan

Test that all of the previous modal's functionality have been retained in the storybook demos.

Old: http://localhost:6006/?path=/docs/components-modal
New: http://localhost:6006/?path=/story/design-system-modal

- [x] The modal can be closed by clicking on the backdrop or pressing Esc
- [x] Right clicking on the backdrop doesn't close the modal
- [x] The modal isn't rendered in the DOM when it's closed
- [x] Other elements that are siblings to the modal have `aria-hidden` set when the modal is open
- [x] Content outside the modal isn't scrollable when it's open
- [x] The modal's contents should be scrollable
- [x] The width/size props should be respected
- [x] Go to "Canvas" mode for the "Modal with standard actions footer" demo and open the modal. In the overflow menu, click on each option and check that they're logged to the "Actions" tab below
- [x] In the "Autofocus" demo, check that the "Address" text field is focused
- [x] In the "Stacking modals" demo, check that you can open a modal on top of another, and switch its props independently via the form within the modal 


## PR Checklist

<!--
Feel free to leave unchecked or remove the lines that are not applicable.
-->

-   [x] Added tests for bugs / new features
-   [x] Updated docs (storybooks, readme)
-   [x] Executed `npm run validate` and made sure no errors / warnings were shown
-   [x] Described changes in `CHANGELOG.md`
-   [ ] Bumped version in `package.json` and `package-lock.json` (`npm --no-git-tag-version version <major|minor|patch>`) [ref](https://docs.npmjs.com/cli/v6/commands/npm-version)
-   [ ] Updated all static build artifacts (`npm run build-all`)

## Versioning

Major

scottlovegrove reviewed (approved):
✅ All works as expected

frankieyan commented:
@scottlovegrove I added a couple more fixes after testing it in Todoist:

* Added a storybook demo to show rendering a modal on top of another
* Added a `Menu` component to the first demo to show menu items can be selected, and that they can be traversed with the keyboard

Could you please have a quick look at the last few commits and give them a spin in Storybook?

* [Add story for modal on top of modal](https://github.com/Doist/reactist/pull/686/commits/affe7e1394e295a5ded674b1384d49da5edcec21)
* [Prevent backdrop click event from propagating](https://github.com/Doist/reactist/pull/686/commits/3dc481838b8eb062e071796041b53b5ddcf78def)
* [Add menu to modal story](https://github.com/Doist/reactist/pull/686/commits/8602515fb3c708e8dbc5dcf0e0e26e58710a3b01)
* [Set default z-index for menu](https://github.com/Doist/reactist/pull/686/commits/94b37cddb26cd84dd69ee8c86b01918a445d9cf5)
* [Allow click events portaled from the modal to be handled](https://github.com/Doist/reactist/pull/686/commits/b4d4d5c464bec2e46ef40eda5b93f9c45aabcf94)
* [Add FocusLock to menu to allow it to trap focus on top of the modal](https://github.com/Doist/reactist/pull/686/commits/2521b52a4724c9cb79f8fc8be32006c31f8820bc)

gnapse commented on src/new-components/modal/modal.stories.mdx:
```diff
@@ -77,10 +82,30 @@ behaviour, see [ARIA: dialog role](https://developer.mozilla.org/en-US/docs/Web/
             <Modal aria-label="Modal with standard actions footer">
                 <ModalHeader>
-                    <Heading level="1">Modal with standard actions footer</Heading>
+                    <Columns>
```
I think this fixes the misaligned modal header:

```suggestion
                    <Columns alignY="center">
```

Although, IMO, this extra button should be put in the header button slot, instead of putting it in the title slot. Something along these lines:

```jsx
<ModalHeader
  button={
    <Columns space="small">
      <Column><CustomButton /></Column>
      <Column><ModalCloseButton /></Column>
    </Columns>
  }
>
  <Heading level="1">Modal with standard actions footer</Heading>
</ModalHeader>
```

gnapse reviewed (approved):
Preliminary review, before I get into the new steps in the test plan:

1. **The header title is not aligned to the close button:**

    <img width="712" alt="CleanShot 2022-08-29 at 11 06 43@2x" src="https://user-images.githubusercontent.com/15199/187233359-59eeb908-6a9e-4352-9c2a-1d3b91df6143.png">

    This only happens in the first story, the one titled "Modal with standard actions footer". So it's most probably a problem with the story, and not with the modal component itself. It probably has to do with how you put this custom menu button in the header. I left a comment in code on what I think may fix this.

    <details>
    <summary>More info</summary>

    I realized while switching between two tabs: one with the new modal, and the other one with the old modal.

    ![CleanShot 2022-08-29 at 11 09 05](https://user-images.githubusercontent.com/15199/187233696-03cdbd47-a9da-42b3-a06e-8f53fa2fd6d8.gif)

    </details>

2. In the "Stacking modal" story, I suggest changing the nested modal to have a different size than the parent modal initially, so that we can see how a modal is floating in top of the other one. That is, a different height and width setting.
